pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,459 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use {
+    crate::{
+        candidate::Candidates,
+        pools::{self, Pool},
+        traits::{ErrAdd, ErrSub},
+        AllTargetPool, Candidate, Config, Delegator, Error, Event, Pallet, PendingOperationKey,
+        PendingOperationQuery, PendingOperations, RequestFilter, Shares, SharesOrStake, Stake,
+        TargetPool,
+    },
+    frame_support::{
+        pallet_prelude::*,
+        traits::{
+            fungible::{Mutate, MutateHold},
+            tokens::{Precision, Preservation},
+            OriginTrait,
+        },
+    },
+    frame_system::pallet_prelude::*,
+    sp_runtime::traits::{CheckedSub, Zero},
+};
+
+pub struct Calls<T>(PhantomData<T>);
+
+impl<T: Config> Calls<T> {
+    pub fn rebalance_hold(
+        _: OriginFor<T>,
```

Comment:
Is this safe? In another call we have

```
        // We don't care about the sender.
        let _ = ensure_signed(origin)?;
```

but here we don't even check if it's signed

Reply:
Not sure if it is necessary to check the origin is signed if we don't care about who it is. We should check in substrate pallet calls that don't care about the sender what they do.

Reply:
I guess it depends if None origins pay fees, or whether we want them to pay fees. I have the impression  that something we might want to do is:

- In the joining/leaving extrinsics, account for fees related to this extrinsics plus the execute-pending-operations ones.
- in the execute-pending-operations extrinsic we might not charge for fees (as it was accounted before)

is rebalance-hold an extrinsic by itself? if so I think we need people that call it to pay fees, because there is no-where where we can account for this right?

Reply:
Can a user call an extrinsic without an origin? The rationale here is that anyone can call this extrinsic, especially in case of slashing to get the funds out of the delegators accounts.

We can have an `ensure_signed` in the `pallet::call` section and don't have an origin parameter in `call::Calls`.

Reply:
> Can a user call an extrinsic without an origin?

Yes, according to @girazoki that's what happens when you press the "send unsigned" button in polkadot js. In that case we need custom logic to handle fees, so I don't think we want to support unsigned transactions here.

Reply:
yes exactly, users will have a clear incentive to do this, hence I would keep it behind a signed origin